service/sharddistributor/executorclient/client.go

```diff
@@ -0,0 +1,118 @@
+package executorclient
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/uuid"
+	"go.uber.org/fx"
+	"go.uber.org/yarpc"
+
+	sharddistributorv1 "github.com/uber/cadence/.gen/proto/sharddistributor/v1"
+	"github.com/uber/cadence/client/sharddistributorexecutor"
+	"github.com/uber/cadence/client/wrappers/errorinjectors"
+	"github.com/uber/cadence/client/wrappers/grpc"
+	"github.com/uber/cadence/client/wrappers/metered"
+	timeoutwrapper "github.com/uber/cadence/client/wrappers/timeout"
+	"github.com/uber/cadence/common/clock"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/dynamicconfig/dynamicproperties"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/rpc"
+	"github.com/uber/cadence/common/service"
+)
+
+//go:generate mockgen -package $GOPACKAGE -source $GOFILE -destination interface_mock.go . ShardProcessorFactory,ShardProcessor,Executor
+
+type ShardProcessor interface {
+	Start(ctx context.Context)
+	Stop()
+	GetShardLoad() float64
+}
+
+type ShardProcessorFactory[SP ShardProcessor] interface {
+	NewShardProcessor(shardID string) (SP, error)
+}
+
+type Executor[SP ShardProcessor] interface {
+	Start(ctx context.Context)
+	Stop()
+
+	GetShardProcess(shardID string) (SP, error)
+}
+
+type Params[SP ShardProcessor] struct {
+	fx.In
+
+	Dispatcher            *yarpc.Dispatcher
+	DC                    *dynamicconfig.Collection
+	MetricsClient         metrics.Client
+	Logger                log.Logger
+	ShardProcessorFactory ShardProcessorFactory[SP]
+	Config                Config
+	TimeSource            clock.TimeSource
+}
+
+func NewExecutor[SP ShardProcessor](params Params[SP]) (Executor[SP], error) {
+	shardDistributorClient, err := createShardDistributorExecutorClient(params.Dispatcher, params.DC, params.MetricsClient, params.Logger)
```

Comment:
A suggestion - could you make it optional? With Yarpcfx, you can get this without these passes.